src/operator/controllers/poduserpassword/db_credentials_pod_reconciler.go

```diff
@@ -79,57 +94,103 @@ func (e *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
 		password, err := createServicePassword()
 		if err != nil {
-			return errors.Wrap(err)
-		}
-		clusterID, err := clusterutils.GetClusterUID(ctx)
-		if err != nil {
-			return errors.Wrap(err)
+			return errors.Wrap(err), ""
 		}
 
-		username := clusterutils.BuildHashedUsername(serviceName, pod.Namespace, clusterID)
-		pgUsername := clusterutils.KubernetesToPostgresName(username)
+		databaseUsername := pod.Annotations[databaseconfigurator.DatabaseUsernameAnnotation]
 
-		secret := buildUserAndPasswordCredentialsSecret(secretName, pod.Namespace, pgUsername, password)
+		secret := buildUserAndPasswordCredentialsSecret(secretName, pod.Namespace, databaseUsername, password)
 		log.WithField("secret", secretName).Debug("Creating new secret with user-password credentials")
 		if err := e.client.Create(ctx, secret); err != nil {
-			return errors.Wrap(err)
+			return errors.Wrap(err), ""
 		}
-		return nil
+		return nil, password
 	}
 
 	if err != nil {
-		return errors.Wrap(err)
+		return errors.Wrap(err), ""
 	}
 	log.Debug("Secret exists, nothing to do")
+	return nil, string(secret.Data["password"])
+}
+
+func (e *Reconciler) ensurePasswordInDatabases(ctx context.Context, pod v1.Pod, password string) error {
+	databases := strings.Split(pod.Annotations[databaseconfigurator.DatabaseAccessAnnotation], ",")
+	username := pod.Annotations[databaseconfigurator.DatabaseUsernameAnnotation]
+	pgServerConfigs := otterizev1alpha3.PostgreSQLServerConfigList{}
+	err := e.client.List(ctx, &pgServerConfigs)
+	if err != nil {
+		return errors.Wrap(err)
+	}
+	mysqlServerConfigs := otterizev1alpha3.MySQLServerConfigList{}
+	err = e.client.List(ctx, &mysqlServerConfigs)
+	if err != nil {
+		return errors.Wrap(err)
+	}
+
+	for _, database := range databases {
+		var dbconfigurator databaseconfigurator.DatabaseConfigurator
+		mysqlConf, found := lo.Find(mysqlServerConfigs.Items, func(config otterizev1alpha3.MySQLServerConfig) bool {
+			return config.Name == database
+		})
+		if found {
+			dbconfigurator, err = mysql.NewMySQLConfigurator(ctx, mysqlConf.Spec)
+			if err != nil {
+				return errors.Wrap(err)
+			}
+			err = dbconfigurator.AlterUserPassword(ctx, username, password)
+			if err != nil {
+				return errors.Wrap(err)
+			}
+			return nil
+		}
+
+		pgServerConf, found := lo.Find(pgServerConfigs.Items, func(config otterizev1alpha3.PostgreSQLServerConfig) bool {
+			return config.Name == database
+		})
+		if found {
+			dbconfigurator, err = postgres.NewPostgresConfigurator(ctx, pgServerConf.Spec)
+			if err != nil {
+				return errors.Wrap(err)
+			}
+			err = dbconfigurator.AlterUserPassword(ctx, username, password)
+			if err != nil {
+				return errors.Wrap(err)
+			}
+			return nil
+		}
+		// We can only reach here if a server config was deleted after the intents operator updated the annotation on the pod
+		// TODO: Log error and continue instead ?
+		return errors.Wrap(fmt.Errorf("no matching server config found for database %s", database))
```

Comment:
consider splitting some of the internal logic to a dedicated function, and utilize the DB configurator interface, to reduce nesting level and duplication:
I'd write something like that:
```
for _, database := range databases {
    dbconfigurator, found, err := e.createDBConfigurator(ctx, database, mysqlServerConfigs.Items, pgServerConfigs.Items)
    if err != nil {...}
    if !found {
        // log error / warning and continue
        continue
    }
    err = dbconfigurator.AlterUserPassword(ctx, username, password)
    if err != nil {
        return errors.Wrap(err)
    }
    return nil
}
```